### PR TITLE
Allow setting maxLoglevel when using the inline client

### PR DESCRIPTION
**Pull request by tomaka**
Close https://github.com/paritytech/substrate-connect/pull/969
cc https://github.com/paritytech/substrate-connect/issues/965

This PR can be reviewed commit by commit.

### CODE REVIEW

**Comment by tomaka** [comment]
Tests are failing because the call to `terminate()` would now be asynchronous.
I don't know whether I should decrease test coverage and increase the code complexity to make them pass.

File: packages/connect/src/connector/smoldot-light.ts [hunk 1]
```diff
@@ -22,26 +22,68 @@ const getStart = () => {
   return startPromise
 }
 
-let clientNumReferences = 0
+const clientReferences: Config[] = [] // Note that this can't be a set, as the same config is added/removed multiple times
 let clientPromise: Promise<Client> | null = null
-const getClientAndIncRef = (): Promise<Client> => {
+let clientReferencesMaxLogLevel = 3
+const getClientAndIncRef = (config: Config): Promise<Client> => {
+  if (config.maxLogLevel && config.maxLogLevel > clientReferencesMaxLogLevel)
```
Comment (wirednkod) [inline packages/connect/src/connector/smoldot-light.ts:29]: I think this works same:
```suggestion
  if (config?.maxLogLevel > clientReferencesMaxLogLevel)
```

File: packages/connect/src/connector/smoldot-light.ts [hunk 2]
```diff
@@ -22,26 +22,68 @@ const getStart = () => {
...
+const decRef = (config: Config) => {
+  const idx = clientReferences.indexOf(config)
+  if (idx === -1) throw new Error("Internal error within smoldot-light")
+  clientReferences.splice(idx, 1)
+
+  // Update `clientReferencesMaxLogLevel`
+  // Note how it is set back to 3 if there is no reference anymore
+  clientReferencesMaxLogLevel = 3
+  for (const cfg of clientReferences.values()) {
+    if (cfg.maxLogLevel && cfg.maxLogLevel > clientReferencesMaxLogLevel)
```
Comment (wirednkod) [inline packages/connect/src/connector/smoldot-light.ts:77]: I think this is the same:
```suggestion
    if (cfg?.maxLogLevel > clientReferencesMaxLogLevel)
```

File: packages/connect/src/connector/smoldot-light.ts [hunk 3]
```diff
@@ -57,21 +99,43 @@ const transformErrors = (thunk: () => void) => {
...
+ */
+export interface Config {
+  /**
+   * The client prints logs in the console. By default, only log levels 1, 2, and 3 (corresponding
+   * respectively to ERROR, WARN, and INFO) are printed.
+   *
+   * In order to more easily debug problems, you can pass 4 (DEBUG) or more.
+   *
+   * This setting is only taken into account between the moment when you use this chain to add a
+   * chain for the first timen, and the moment when all the chains that you have added have been
```
Comment (wirednkod) [inline packages/connect/src/connector/smoldot-light.ts:113]:
```suggestion
   * chain for the first time, and the moment when all the chains that you have added have been
```

**Review by wirednkod — APPROVED** [review]
minor things and a typo - overall lgtm

File: packages/connect/src/connector/smoldot-light.ts (hunk 1 above)
Comment (tomaka) [reply packages/connect/src/connector/smoldot-light.ts:29]: I don't really like this change. You're comparing null with a number, which doesn't make sense and should be forbidden by TypeScript if it was me.

Comment (wirednkod) [reply packages/connect/src/connector/smoldot-light.ts:29]: That is not a comparison of null with a number. The `if (config?.maxLogLevel > clientReferencesMaxLogLevel)` actually translates to `if ((config === null || config === void 0 ? void 0 : config.maxLogLevel) > clientReferencesMaxLogLevel)` so its actual a comparison between numbers, but its just a detail and doesn't really matter which one we use. Result is same

**Review by wirednkod — APPROVED** [review]

**Pull request merged**